libfreerdp/codec/h264_ffmpeg.c

```diff
@@ -197,16 +209,54 @@ static int libavcodec_decompress(H264_CONTEXT* h264, const BYTE* pSrcData,
 
 	if (status < 0)
 	{
-		WLog_ERR(TAG, "Failed to decode video frame (status=%d)", status);
+		WLog_ERR(TAG, "Failed to decode video frame (%s)", av_err2str(status));
 		return -1;
 	}
 
+#ifdef WITH_VAAPI
+	sys->videoFrame->format = AV_PIX_FMT_YUV420P;
+
+	while (TRUE)
+	{
+		status = avcodec_receive_frame(sys->codecDecoderContext, sys->hwVideoFrame);
+
+		if (status == AVERROR(EAGAIN))
+		{
+			continue;
+		}
+		else if (status < 0)
+		{
+			WLog_ERR(TAG, "Error decoding frame from hardware: %s", av_err2str(status));
+		}
+
+		if (sys->hwVideoFrame->format == sys->hw_pix_fmt)
+		{
+			sys->videoFrame->width = sys->hwVideoFrame->width;
+			sys->videoFrame->height = sys->hwVideoFrame->height;
+			status = av_hwframe_transfer_data(sys->videoFrame, sys->hwVideoFrame, 0);
+
+			if (status < 0)
+			{
+				WLog_ERR(TAG, "Error transferring frame from hardware: %s", av_err2str(status));
+			}
+		}
+		else
+		{
+			status = av_frame_copy(sys->videoFrame, sys->hwVideoFrame);
+		}
+
+		break;
+	}
+
+#else
+
 	do
```

Comment:
Please keep the original loop add #ifdefs statement before and another after for the VAAPI case